coverage.c

```diff
@@ -0,0 +1,632 @@
+/* coverage.c -- samtools coverage subcommand
+ 
+   Author: Florian P Breitwieser <florian.bw@gmail.com>
+
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.  */
+
+/* This program calculates coverage from multiple BAMs
+ * simutaneously, to achieve random access and to use the BED interface.
+ * To compile this program separately, you may:
+ *
+ *   gcc -g -O2 -Wall -o bamcov -D_MAIN_BAMCOV coverage.c -lhts -lz
+ */
+
+// C headers
+#include <config.h>
+
+#include <ctype.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>  // variadic functions
+#include <limits.h>  // INT_MAX
+#include <math.h>    // round
+#include <stdbool.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
```

Comment:
`ioctl.h` is not included in the MSYS2 toolchain, so it will not build on Windows.
Appveyor doesn't flag this, because it builds on `develop` for pull requests (there is still some work to be done on the config script).

Reply:
Thanks for catching this. I use it to get terminal width, when the width of the histogram is the whole width. I'll check for alternatives on Windows.

Reply:
The environment variable `COLUMNS`, when it's present, otherwise default to 80.

Reply:
Just tried with `getenv("COLUMNS")`. In principle that would be a great solution, but it seems that I need to call `export COLUMNS` in the shell for the `COLUMNS` to be set in the program. Is there a workaround for that?

Reply:
> The process that is spawned from the shell won't see any variables that are not exported, and at least OSX bash does not export COLUMNS by default.

Huh. You're right. That's odd. I learned something.

```console
$ echo $COLUMNS
80
$ env | grep COLUMNS
$
```

Reply:
Me too — I always assumed $LINES/$COLUMNS were exported too. Back to `ioctl` with appropriate ifdefs I guess!

Reply:
Just setup a MSYS2 environment to test; for me it compiles and works with the ioctl header. Are the instructions on how to build samtools w/ MSYS2? Also tried GetConsoleScreenBufferInfo w/ MSYS2, but doesn't give me the correct values so far.

I'm not cross-compiling usually, so excuse my ignorance: `_WIN32` is not defined when compiling with gcc in MSYS2. How should I check the system?

Reply:
I should have added a few more details. MSYS2 comes in two flavours:

- the MSYS2 line, which is a fork of Cygwin, is more POSIX-like and links all programs with a compatibility library (`msys-2.0.dll`). This is what you have probably installed. It usually ships **gcc 7** and the `msys2-runtime-devel` package, which includes `ioctl.h` (`pacman -Fs ioctl.h`).
- the MINGW line provides the `mingw-w64-{i686,x86_64}-toolchain`, which includes **gcc 8** and links with the native Windows libraries. And this is what Appveyor uses. You can see that at line 13 in the build [log](https://ci.appveyor.com/project/valeriuo/samtools-dp6yo/builds/22378373):

`sh -lc "pacman -S --noconfirm --needed base-devel mingw-w64-x86_64-toolchain mingw-w64-x86_64-zlib mingw-w64-x86_64-bzip2 mingw-w64-x86_64-xz mingw-w64-x86_64-curl"`

`GetConsoleScreenBufferInfo` seems to be working only with `cmd` consoles. 

Anyway, the `#include` and the code have to be guarded by `#if defined(__MSYS__) || defined (__CYGWIN__)`. 
`_WIN32` comes with the MINGW toolchain.

Reply:
Thanks a lot for the details @valeriuo ! That `GetConsoleScreenBufferInfo` only works with `cmd` consoles complicates it, but the feature is anyway not the most important - histogram width can still be defined manually. I'll submit a new version.

Reply:
What's the minimum GCC version that should be supported? What do you think of `#if defined(__has_include) && __has_include(<sys/ioctl.h>)` as test? `__has_include` was added in GCC 5.

Reply:
Using `__has_include` would be trying too hard. (But see the macros in HTSlib's _hts_defs.h_ for similar things that **are** worthwhile.)

You could use `configure` to see if ioctl was available, but that would be trying too hard too.

You mostly just want it to **build** on windows. So `#ifndef __WINDOWS_WHATEVER_` use ioctl `#else` just use `$COLUMNS` or (`#if WINDOWS`) `GetConsoleScreenBufferInfo` or a hardcoded 80.

Reply:
Sounds good. Just pushed a version that checks with `__has_include` if it's available, and check w/ ifdefs otherwise. Can refactor that to not use `__has_include`. In the current version, `getenv("COLUMNS")` is checked first, then ioctl if it has it or windows GetConsoleScreenBufferInfo. Might be better to check for ioctl/GetConsoleScreenBufferInfo first, and do getenv if those are not available.

I also changed the options to have full UTF mode (with eight block characters) and the legacy mode (with two block characters) on a more equal footing: `samtools coverage -m` and `samtools coverage --histogram` show legacy mode, `samtools coverage -M` and `samtools coverage --Histogram` show the full UTF set. Also switched to using full terminal width for the histogram by default. I think that's desirable but am open to discussion.

Reply:
For the option names, I suggest…
```
samtools coverage -m
samtools coverage --histogram
samtools coverage -m --ascii
samtools coverage --histogram --ascii
```

Reply:
An environment variable would be helpful to set the default.
```sh
export SAMTOOLS_ASCII=1
```
in which case there should also be a `--utf8` option to override the environment variable.

Reply:
I think the `__has_include` is a bit over the top.  Usually for this sort of thing we just check for the normal Windows defines, and otherwise assume the normal unix-y stuff is available.

`TIOCGWINSZ` seems fairly universal, but as ioctls are somewhat system-dependent it might be worth putting in an `#ifdef TIOCGWINSZ` check in to ensure it's really there.

Reply:
I'm not sure on using an environment variable to control this behaviour.  I'm not aware of other subcommands where command line options are alternatively given via the environment instead.

Not saying it's necessarily a bad idea, but if we do it I think we should be consistent and think it through rather than winging it for this one sub-command, eg with sensible name spaces such as SAMTOOLS_COVERAGE_OPTS=-m or similar, and permit it to simply be an extension to argc/argv.

Reply:
For reproducibility, I'm against allowing environment variables to affect behaviour other than simple presentation issues. One advantage of `SAMTOOLS_ASCII` over say `SAMTOOLS_COVERAGE_OPTS` is that it could be used by other components of `samtools` other than `coverage` that run into the same UTF/ASCII issue.